pkg/io/option.go

```diff
@@ -0,0 +1,94 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package io
+
+import (
+	"cloud.google.com/go/storage"
+	"gocloud.dev/blob"
+	"google.golang.org/api/googleapi"
+)
+
+// A WriterOption is an option for the opener Writer method
+type WriterOption interface {
+	Apply(*storage.Writer, *blob.WriterOptions)
+}
+
+// WithBufferSize returns a WriterOption that specifies the chunk / buffer size
+func WithBufferSize(size int64) WriterOption {
```

Comment:
This is a question of taste, but I personally prefer the simple option of having a public struct with pointer fields for all options we allow to set that implements `WriterOptions` and then make the `WithSomething` funcs use that struct, because 

* it reduces the amount of types and `WriterOption` implementations that are needed
* It allows to see all possible options quickly by jumping into the definition of that struct, rather than having to jump into all `With` funcs

```
type BlobOption struct {
  Size *int64
  ContentType *string
  ...
}

func (bo *BlobOption) Apply(writer *storage.Writer, o *blob.WriterOptions) {
  if writer != nil {
    if bo.Size != nil {
      writer.ChunkSize = bo.Size
   }
   if bo.ContentType != nil {
      writer.ContentType = bo.ContentType
   }
  }
  if o != nil {
      ...
   }
}

func WithBufferSize (size int64) WriterOption {
   return &BlobOption{Size: utilpointer.Int64(size)}
}
```

Reply:
(Having one public type rather than many `WithSomething` funcs based on distinct private types would also allow us to use that type in the places where we currently use `storage.ObjectAttrs`, e.G. in `prow/pod-utils/gcs/upload.go`)

Reply:
Thx, you're right. Your variant is better